ceph/ceph
# mgr/dashboard: support creating OSDs on spare devices

On OSD page, a form is added to allow creating OSDs from non-occupied
devices. User can:
- use filters to select some primary devices for OSD.
- use filters to select WAL/DB devices as shared devices if needed.

Note: This feature requires orchestrator support.

Frontend changes:
- Extract inventory devices component from inventory. We need to reuse
it.
- The inventory devices component supports column filters. The available
options for filters are determined by all possible values in a column.
- Add a button on OSD list page to display OSD creation form.
- Add OSD creation form, it allows selecting primary/WAL/DB devices for
OSDs.
- Add a preview modal to preview OSD creation. Currently, the Drive
Group Specification is displayed. This feature will be completed when we
have library to preview OSD creation.

Fixes: https://tracker.ceph.com/issues/40335
Fixes: https://tracker.ceph.com/issues/42076

Signed-off-by: Kiefer Chang <kiefer.chang@suse.com>

## Depends
- [x] #30662 (Device attributes are changed, should re-arrange Inventory Devices columns after this PR)
- [x] #31189 (Add OSDs column in Inventory Devices table)

## Checklist


## How to test

Currently, no orchestrator backend fully supports drive group. We can use `test_orchestrator` backend and test the feature by feeding some fake data.

```
cd /ceph/build
bin/ceph mgr module enable test_orchestrator
bin/ceph orchestrator set backend test_orchestrator
wget https://raw.githubusercontent.com/bk201/ceph-resources/master/issues/40335/dummy.json -O /tmp/dummy.json
bin/ceph test_orchestrator load_data -i /tmp/dummy.json
```
 
---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`

</details>
